### PR TITLE
Feature/sax

### PR DESCRIPTION
**Pull request by dhiguero**
#### What's this PR do?

Adds the SAX-VSM algorithm to the SOLMA library. This algorithm permits to transform a stream of Doubles into a stream of words. The stream of words is then compared using the SAXDictionary as to produce a prediction of which class the last window of words belongs to.

References on the algorithm can be found at:

```
Senin, Pavel, and Sergey Malinchik. "Sax-vsm: Interpretable time series classification using
SAX and vector space model." Data Mining (ICDM), 2013 IEEE 13th International Conference on.
IEEE, 2013.
```

#### Where should the reviewer start?

Both the `SAX` and `SAXDictionary` classes contain the entry points for the two elements required in the implementation of SAX-VSM. The `SAX` class contains the implementation of the SAX algorithm which is trained using a `Dataset[Stream]` in order to determine statistics on the data set distribution. Once fitted, the algorithm will transform a `DataStream[Double]` into a `DataStream[String]` with the resulting words.

The `SAXDictionary` takes `DataSet[(String, Int)]` for defining the classes in the dictionary during the training phase. Once trained, the dictionary builds a TF-IDF matrix that will be used to compare a set of incoming windows with the existing classes. For that the algorithm accept `DataStream[String]` and returns a `DataStream[SAXPrediction]` with the identifier and distance to the nearest class.

#### What is missing?

As it is not clear what would be the number of letters in the alphabet that is optimal for the proteus dataset, it remains for future developments to increase the number of supported symbols in the `Cuts` class.

#### How should this be manually tested?

Unit tests are included
#### Questions:

### CODE REVIEW

File: src/main/scala/eu/proteus/solma/fd/FrequentDirections.scala [hunk 1]
```diff
@@ -89,10 +89,10 @@ object FrequentDirections {
   implicit def fitNoOp[T] = {
     new StreamFitOperation[FrequentDirections, T]{
       override def fit(
-          instance: FrequentDirections,
```
Comment (VenturaDelMonte) [inline src/main/scala/eu/proteus/solma/fd/FrequentDirections.scala]: Why did you change this file? Are you fixing the indentation?

File: src/main/scala/eu/proteus/solma/moments/MomentsEstimator.scala [hunk 2]
```diff
@@ -101,10 +101,10 @@ object MomentsEstimator {
   implicit def fitNoOp[T] = {
```
Comment (VenturaDelMonte) [inline src/main/scala/eu/proteus/solma/moments/MomentsEstimator.scala]: Why did you change this file? Are you fixing the indentation?

File: src/main/scala/eu/proteus/solma/sax/Cuts.scala [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+  /**
+   * Get the distance matrix resulting from splitting the normal curve into n cuts that form
+   * a set of ranges. The matrix contains the distance between the ranges.
+   * @param numberCuts The number of cuts.
+   * @return An matrix of doubles.
+   */
+  def distanceMatrix(numberCuts: Int) : Array[Array[Double]] = {
+    numberCuts match {
+      case 2 => Cuts.Distances2
```
Comment (VenturaDelMonte) [inline src/main/scala/eu/proteus/solma/sax/Cuts.scala]: Would an UnsupportedOperationException fit better in this case?

File: src/main/scala/eu/proteus/solma/sax/SAXDictionaryFitOperation.scala [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    val fitting = input.map(_ match {
+      case (value : String, classId : String) => (value.toString, classId.toString)
+      case (value : String, classId : Int) => (value.toString, classId.toString)
+      case _ => throw new IllegalArgumentException("Expecting tuples of (String, String) or (String, Int)")
+    }).groupBy(1)
+      .reduceGroup[WordBag](wordBagReduceFunction)
+      .reduceGroup[BagDictionary](toDictionary)
+
```
Comment (VenturaDelMonte) [inline src/main/scala/eu/proteus/solma/sax/SAXDictionaryFitOperation.scala]: The collect operation gathers everything on one machine. How large is the resulting dictionary?

File: src/main/scala/eu/proteus/solma/sax/SAXDictionaryPredictOperation.scala [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        input.foreach(w => {
+          freq.putIfAbsent(w.toString, new AtomicLong(0))
+          freq.get(w.toString).incrementAndGet()
+        })
+
+        val prediction = instance.dictionary.get.predict(freq)
+        out.collect(prediction)
+      }
+    }
+
```
Comment (VenturaDelMonte) [inline src/main/scala/eu/proteus/solma/sax/SAXDictionaryPredictOperation.scala]: An AllWindowedStream is executed with parallelism=1, thus I was wondering whether predictions can be served on parallel streams (KeyedStream).

File: src/main/scala/eu/proteus/solma/sax/SAXStreamTransformOperation.scala [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+    val avg = instance.trainingAvg
+    val std = instance.trainingStd
+    if(avg.isEmpty || std.isEmpty){
+      throw new RuntimeException("You must train the SAX before calling transform")
+    }
+
+    val wordSize = instance.getWordSize()
+    val paaFragmentSize = instance.getPAAFragmentSize()
+
```
Comment (VenturaDelMonte) [inline src/main/scala/eu/proteus/solma/sax/SAXStreamTransformOperation.scala]: Same question as above: could this be applied on a keyed stream?

File: src/main/scala/eu/proteus/solma/utils/FlinkSolmaUtils.scala [hunk 7]
```diff
@@ -20,6 +20,7 @@ package eu.proteus.solma.utils
```
Comment (VenturaDelMonte) [inline src/main/scala/eu/proteus/solma/utils/FlinkSolmaUtils.scala]: What's the purpose of this change?

File: src/main/scala/eu/proteus/solma/fd/FrequentDirections.scala (hunk 1 above)
Comment (dhiguero) [inline src/main/scala/eu/proteus/solma/fd/FrequentDirections.scala]: I did make changes locally, and then reverted them back after the discussion about the fit for datasets. I will revert those changes so it matches the original version.

File: pom.xml [hunk 8]
```diff
@@ -138,6 +138,22 @@
 		</dependency>
 
 		<dependency>
```
Comment (VenturaDelMonte) [inline pom.xml]: I missed this one last time I checked, but flink-ml is already imported at line 108.

File: src/main/scala/eu/proteus/solma/sax/SAXDictionaryPredictOperation.scala [hunk 9]
```diff
@@ -42,28 +45,31 @@ class SAXDictionaryPredictOperation[T] extends PredictDataStreamOperation[SAXDic
...
         window: GlobalWindow,
-        input: Iterable[T],
+        input: Iterable[(T, Int)],
         out: Collector[SAXPrediction]): Unit = {
 
         val freq = new JHashMap[String, AtomicLong]
         input.foreach(w => {
-          freq.putIfAbsent(w.toString, new AtomicLong(0))
-          freq.get(w.toString).incrementAndGet()
+          freq.putIfAbsent(w._1, new AtomicLong(0))
```
Comment (VenturaDelMonte) [inline src/main/scala/eu/proteus/solma/sax/SAXDictionaryPredictOperation.scala]: Why AtomicLong? Since the HashMap is only local, there will be no concurrent access, therefore, i do not think that an atomic increment is needed at all. Or, am i missing anything?

**Comment by dhiguero** [comment]
I have modified the code to not use AtomicLong as it is not necessary for this use case. With regards to the use of StreamEvent, I think it is a change that can be done in a future PR if required.

**Comment by VenturaDelMonte** [comment]
Merged.
